libraries/AP_Compass/CompassCalibrator.cpp

```diff
@@ -69,145 +70,154 @@
+    cal_settings.fix_orientation = fix_orientation;
 }
 
 void CompassCalibrator::start(bool retry, float delay, uint16_t offset_max, uint8_t compass_idx)
 {
-    if (running()) {
+    if (compass_idx > COMPASS_MAX_INSTANCES) {
         return;
     }
-    _offset_max = offset_max;
-    _attempt = 1;
-    _retry = retry;
-    _delay_start_sec = delay;
-    _start_time_ms = AP_HAL::millis();
-    _compass_idx = compass_idx;
-    set_status(Status::WAITING_TO_START);
-}
 
-void CompassCalibrator::get_calibration(Vector3f &offsets, Vector3f &diagonals, Vector3f &offdiagonals, float &scale_factor)
-{
-    if (_status != Status::SUCCESS) {
+    WITH_SEMAPHORE(state_sem);
+    // Don't do this while we are already started
+    if (cal_state.status == Status::RUNNING_STEP_ONE || 
+       cal_state.status == Status::RUNNING_STEP_TWO) {
         return;
     }
+    cal_settings.offset_max = offset_max;
+    cal_settings.attempt = 1;
+    cal_settings.retry = retry;
+    cal_settings.delay_start_sec = delay;
+    cal_settings.start_time_ms = AP_HAL::millis();
+    cal_settings.compass_idx = compass_idx;
+
+    if (_cal[compass_idx] == nullptr) {
+        _cal[compass_idx] = this;
+    }
 
-    offsets = _params.offset;
-    diagonals = _params.diag;
-    offdiagonals = _params.offdiag;
-    scale_factor = _params.scale_factor;
+    if (!_thread_started) {
+        if (!hal.scheduler->thread_create(FUNCTOR_BIND(this, &CompassCalibrator::update_trampoline, void), "compasscal", 1024, AP_HAL::Scheduler::PRIORITY_IO, 0)) {
+            gcs().send_text(MAV_SEVERITY_CRITICAL, "CompassCalibrator: Cannot start compass thread.");
+            return;
+        }
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "CompassCalibrator: Thread Initialised");
+        _thread_started = true;
+    }
+
+    // Request status change to Waiting to start
+    _requested_status = Status::WAITING_TO_START;
+    _status_set_requested = true;
 }
 
-float CompassCalibrator::get_completion_percent() const
+// Record point mag sample and associated attitude sample to intermediate struct
+void CompassCalibrator::new_sample(const Vector3f& sample)
 {
-    // first sampling step is 1/3rd of the progress bar
-    // never return more than 99% unless _status is Status::SUCCESS
-    switch (_status) {
-        case Status::NOT_STARTED:
-        case Status::WAITING_TO_START:
-            return 0.0f;
-        case Status::RUNNING_STEP_ONE:
-            return 33.3f * _samples_collected/COMPASS_CAL_NUM_SAMPLES;
-        case Status::RUNNING_STEP_TWO:
-            return 33.3f + 65.7f*((float)(_samples_collected-_samples_thinned)/(COMPASS_CAL_NUM_SAMPLES-_samples_thinned));
-        case Status::SUCCESS:
-            return 100.0f;
-        case Status::FAILED:
-        case Status::BAD_ORIENTATION:
-        case Status::BAD_RADIUS:
-            return 0.0f;
-    };
-    // will not get here if the compiler is doing its job (no default clause)
-    return 0.0f;
+    WITH_SEMAPHORE(sample_sem);
```

Comment:
I disagree, we should use blocking, but make sure the sample_sem is only held for a short time in the thread

Reply:
Why? If we can't push a sample from the sensor thread it isn't a big deal because we can come back around and provide it on the next loop. At a minimum we would want it to be a short time period on the block, rather then the unlimited one.

Reply:
The only other place this sem is used is when pulling the sample, which is just copying the sample, after I make the change to give the semaphore before accept_sample, it should be ok afterwards.

Reply:
non-blocking calls can end up with timing so that it never gets through (we've seen that before). If the other end is a simple copy then we should do it blocking